src/app/Library/CrudPanel/Traits/MorphRelationships.php

```diff
@@ -0,0 +1,210 @@
+        }
+
+        return $input;
+    }
+
+    /**
+     * This function created the MorphTo relation fields in the CrudPanel.
+     *
+     * @param  array  $field
+     * @return void
+     */
+    private function createMorphToRelationFields(array $field, $morphTypeFieldName, $morphIdFieldName)
+    {
+        $morphTypeField = static::getMorphTypeFieldStructure($field['name'], $morphTypeFieldName);
+        $morphIdField = static::getMorphIdFieldStructure($field['name'], $morphIdFieldName, $morphTypeFieldName);
+        $morphIdField['morphMap'] = $morphTypeField['morphMap'] = (new $this->model)->{$field['name']}()->morphMap();
+        $field['subfields'] = [$morphTypeField, $morphIdField];
+
+        return $field;
+    }
+
+    /**
+     * Return the relation field names for a morphTo field.
+     *
+     * @param  string  $relationName  the morphto relation name
+     * @return array
+     */
+    private function getMorphToFieldNames(string $relationName)
+    {
+        $relation = (new $this->model)->{$relationName}();
+
+        return [$relation->getMorphType(), $relation->getForeignKeyName()];
+    }
+
+    /**
+     * Make sure morph fields have the correct structure.
+     *
+     * @param  array  $field
+     * @return array
+     */
+    private function makeSureMorphSubfieldsAreDefined(array $field)
+    {
+        if (isset($field['relation_type']) && $field['relation_type'] === 'MorphTo') {
+            [$morphTypeFieldName, $morphIdFieldName] = $this->getMorphToFieldNames($field['name']);
+            if (! $this->hasFieldWhere('name', $morphTypeFieldName) || ! $this->hasFieldWhere('name', $morphIdFieldName)) {
+                // create the morph fields in the crud panel
+                $field = $this->createMorphToRelationFields($field, $morphTypeFieldName, $morphIdFieldName);
+                foreach ($field['morphOptions'] ?? [] as $morphOption) {
+                    [$key, $label, $options] = $this->getMorphOptionStructured($morphOption);
+                    $field = $this->addMorphOption($field, $key, $label, $options);
+                }
+            }
+        }
+
+        return $field;
+    }
+
+    /**
+     * This function is responsible for setting up the morph fields structure.
+     * Developer can define the morph structure as follows:
+     *  'morphOptions => [
+     *       ['nameOnAMorphMap', 'label', [options]],
+     *       ['App\Models\Model'], // display the name of the model
+     *       ['App\Models\Model', 'label', ['data_source' => backpack_url('smt')]
+     *  ]
+     * OR
+     * ->addMorphOption('App\Models\Model', 'label', ['data_source' => backpack_url('smt')]).
+     *
+     * @param  string|array  $fieldName
+     * @param  string  $key
+     * @param  string|null  $label
+     * @param  array  $options
+     * @return void|array
+     */
+    public function addMorphOption($fieldName, string $key, $label = null, array $options = [])
```

Comment:
This method is 50 lines long, and does a lot of stuff. So let's please add comments before each thing we do.